File: src/exemplos/04_Subalgoritmos/04-bhaskara.py
#  -*- coding: utf-8 -*-
#    @package: 04-bhaskara.py
#     @author: Guilherme N. Ramos (gnramos@unb.br)
# @disciplina: Algoritmos e Programação de Computadores
#
# Exemplo de implementação de funções.


def leia_real():
    '''Lê um valor real.'''
    return float(input('Digite o número: '))


def absoluto(n):
    '''Retorna o valor absoluto do número dado.'''
    if n < 0:
        return -n
    return n


def raiz(n):
    '''Calcula a raiz quadrada do número (pelo método Babilônico).'''
    r = n / 2.0

    while (absoluto(r*r - n) > r):
        r = (r+(n/r))/2

    return r


def delta(a, b, c):
    '''Calcula o Delta.'''
    return (b*b) - (4*a*c)


a = leia_real()
b = leia_real()
c = leia_real()

r1 = (-b + raiz(delta(a, b, c))) / 2
r2 = (-b - raiz(delta(a, b, c))) / 2

print('As raízes são: {} e {}'.format(r1, r2))

File: src/exemplos/04_Subalgoritmos/04-bhaskara.c
/**      @file: 04-bhaskara.c
 *     @author: Guilherme N. Ramos (gnramos@unb.br)
 * @disciplina: Algoritmos e Programação de Computadores
 *
 * Exemplo de implementação de funções. */

#include <stdio.h>

/* Lê um valor real. */
float leia_real() {
    float num;

    printf("\nDigite o número: ");
    scanf("%f", &num);

    return num;
}

/* Retorna o valor absoluto do número dado. */
float absoluto(float n) {
    return (n < 0 ?  -n : n);
}

/* Calcula a raiz quadrada do número (pelo método Babilônico). */
float raiz(float n) {
    float r = n / 2;

    while(absoluto(r*r - n) > r)
        r = (r+(n/r))/2;

    return r;
}

/* Calcula o Delta. */
float delta(float a, float b, float c) {
    return (b*b) - (4*a*c);
}

int main() {
    float a, b, c, r1, r2;

    a = leia_real();
    b = leia_real();
    c = leia_real();

    r1 = (-b + raiz(delta(a, b, c))) / 2;
    r2 = (-b - raiz(delta(a, b, c))) / 2;

    printf("As raízes são: %f e %f\n", r1, r2);

    return 0;
}
